# Insert Title of the Case Study

Submitted by:   
**Lorraine Renee B. Cortel**  
**Dennis H. Lu**  
**Danielle Kirsten T. Sison**

**CSMODEL – S14**

Submitted to:   
**Mr. Thomas James Z. Tiam-Lee**

Date of Submission:  
**September 22, 2020**

***

## 1. Data Description

The dataset used in this case study is based on the medical records of heart failure patients collected at the Faisalabad Institute of Cardiology and at the Allied Hospital in Faisalabad (Punjab, Pakistan) from April 2015 to December 2015 [1, 2]. The researchers of the original study ensured that all patients who participated in the data collection are aware of the situation and have provided their consent [2]. Additionally, the original study was approved by Institutional Review Board of Government College University, Faisalabad-Pakistan and followed the principles of Helsinki Declaration [2].

The dataset consisted of 299 observations, wherein one observation is a medical report of a heart failure patient. In total, there are 105 female patients and 194 male patients whose medical report was included in the dataset. All of the patients are above 40 years old and have left ventricular systolic dysfunction (LVSD), a complication caused when the left ventricle of the heart loses its ability to contract. The left ventricle of the heart is responsible for pumping oxygen-rich blood into circulation and LVSD results to the heart having weaker pumping force. This puts all the patients for this dataset under Class III or IV of New York Heart Association's (NYHA) Functional Classification – a classification for how limited a patient is during physical activity. Patients under Class III and IV are vulnerable to fatigue easily and experiences discomfort when any physical activity is performed. 

Based on Ahmad et. al's case study, the following features were considered as potential variables to explain mortality caused by cardiovascular diseases like heart failure: age, gender, blood pressure (BP), ejection fraction (EF), creatinine phosphokinase (CPK), platelet count, serum sodium, serum creatinine, and whether the patient has anemia, diabetes, and is a smoker [2]. 

Some information like anemia, high blood pressure, diabetes, sex, and smoking where represented as binary data – with 1 as an indicator that the feature is true for that particular patient and 0 amounting to false. For example, a patient who has 1 on anemia but 0 for high blood pressure means that the patient has anemia but no symptoms for high blood pressure. Anemia classifications were determined by the hospital physician using the patient's haematocrit level. If a patient has haematocrit levels less than 36%, the minimum for a normal haematocrit level, then they were considered as anemic. For the classification of high blood pressure and smoking status, the original researchers used physician reports. 

Numerical risk factors were generated from the patients' blood reports and medical records. EF pertains to the amount of blood pumped out by the left ventricle for each contraction and is represented in percentage format. CPK describes the level of CPK enzymes in the blood and is represented in micrograms per liter (mcg/L). CPK was used as a risk factor for high levels of CPK in the blood could indicate heart failure or injury because more CPK is flowing in the blood during muscle tissue damage. A risk factor related with CPK is serum creatinine. Serum creatinine is a waste product generated by creatinine that comes from the normal wear and tear on muscles of the body. Like CPK, high levels of serum creatinine could indicate heart failure. In the dataset, serum creatinine is represented in values of milligrams per deciliter (mg/dL). Serum sodium is the indicator used to measure the sodium level in a patient's blood. It is represented as milliequivalents per litre (mEq/L). Sodium is important for ensuring that muscles and nerves are functioning correctly. Therefore, abnormally low levels of sodium in the blood might be caused by a heart failure. Platelet count was recorded as kiloplatelets per millilitre (kiloplatelets/mL).

Other features of the dataset includes a follow-up period measured in days and a variable for whether a patient is still alive or not called "death event". Like the rest of binary data in the study, 1 indicates that a patient is already dead, while 0 is the indicator for patients that are still alive. Follow-up period refers to the act of monitoring a patient's health after treatment, including when a patient participates in a clinical study or trial for a period of time. In the study, follow-up time varies per patient and is the basis for the death event feature, meaning patients that died within the follow-up period will have their death event recorded as 1.

Regardless of the field of study, accurate data collection is important to maintain the integrity of the research. Appropriate selection of data collection instruments and clearly outlined instructions for the actual data gathering procedure reduces the likelihood of errors occuring. Thus, the dataset used in the original case study and this case study should not be an exception. Failure to comply to these methods could possibly result to having difficulties in answering the proposed research questions. Furthermore, even if the data analysis was executed, there is no guarantee that acquired results are accurate and clearly reflects the initial intention of the whole data gathering process. 

## 2. Exploratory Data Analysis

### 2.1. Dataset Description and Variables

In this case study, a dataset called `heart_failure_clinical_records_dataset` is used for data analysis and data modelling. Data modelling is based on the formulated research questions made by the authors of this case study. The aim of the research questions is to aid the authors in creating a comprehensive analysis of the dataset to extract insights, answer the formulated questions, and draw conclusions about the contents of the dataset.

The dataset is available as a **.csv** file under the folder `dataset`.

As mentioned before, the dataset contains medical records of patients that have heart failure. The dataset consists of **299 observantions**, meaning that there is a total of 299 patients that participated in the data gathering, and these observations are represented as rows in the Excel file. The columns in the Excel file dennotes the variables of the dataset. In total, there are **13 variables** displayed.

The description of each variable in the dataset is provided below.
* **`age`:** Refers to the age of the patient.
* **`anaemia`:** Indicates whether a patient has anemia (1) or not (0).
* **`creatinine_phosphokinase`:** Refers to the amount of the CPK enzyme in the blood – measured in mcg/L.
* **`diabetes`:** Indicates whether a patient has diabetes (1) or not (0).
* **`ejection_fraction`:** Refers to the percentage of blood that leaves the heart at each contraction.
* **`high_blood_pressure`:** Indicates whether a patient has high blood pressure (1) or not (0).
* **`platelets`:** Refers to the platelet count in blood – measured in kiloplatelets/mL.
* **`serum_creatinine`:** Refers to the amount of creatinine in the blood – measured in mg/dL.
* **`serum_sodium`:** Refers to the amount of sodium in the blood – measured in mEq/L.
* **`sex`:** Indicates whether a patient is male (1) or female (0).
* **`smoking`:** Indicates whether a patient is a smoker (1) or not (0).
* **`time`:** Refers to the number of days for the follow-up period of each patient.
* **`DEATH_EVENT`:** Indicates whether the patient is died during the follow-up period (1) or is still alive (0).

### 2.2. Importing Libraries

To properly visualize the dataset, libraries such `numpy` and `pandas`, and function collections like `matplotlib.pyplot` has to be imported.

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

### 2.3. Loading the Dataset

Displaying the dataset requires the use of `pandas`. By implementing the functions `read_csv`, the dataset can be loaded as a pandas DataFrame.

In [2]:
heartfailure_df = pd.read_csv("dataset/heart_failure_clinical_records_dataset.csv")

To check whether the variables described before coincides with the actual content of the loaded dataset file, the `info` function will be called. The `info` function also displays general information about the dataset including the index range count and column count. It also displays information about the number and name of the columns, non-null data count per column, and the data type of observations in a column. 

In [3]:
heartfailure_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 299 entries, 0 to 298
Data columns (total 13 columns):
age                         299 non-null float64
anaemia                     299 non-null int64
creatinine_phosphokinase    299 non-null int64
diabetes                    299 non-null int64
ejection_fraction           299 non-null int64
high_blood_pressure         299 non-null int64
platelets                   299 non-null float64
serum_creatinine            299 non-null float64
serum_sodium                299 non-null int64
sex                         299 non-null int64
smoking                     299 non-null int64
time                        299 non-null int64
DEATH_EVENT                 299 non-null int64
dtypes: float64(3), int64(10)
memory usage: 30.5 KB


To provide a preview of the dataset, the first 5 observations will be displayed below.

In [4]:
heartfailure_df.head()

age  anaemia  creatinine_phosphokinase  diabetes  ejection_fraction  \
0  75.0        0                       582         0                 20   
1  55.0        0                      7861         0                 38   
2  65.0        0                       146         0                 20   
3  50.0        1                       111         0                 20   
4  65.0        1                       160         1                 20   

   high_blood_pressure  platelets  serum_creatinine  serum_sodium  sex  \
0                    1  265000.00               1.9           130    1   
1                    0  263358.03               1.1           136    1   
2                    0  162000.00               1.3           129    1   
3                    0  210000.00               1.9           137    1   
4                    0  327000.00               2.7           116    0   

   smoking  time  DEATH_EVENT  
0        0     4            1  
1        0     6            1  
2        1     7            1  
3        0     7            1  
4        0     8            1

### 2.4. Data Cleaning

Data cleaning needs to be conducted to assure that the insights and conclusions drawn from the analysis of the dataset are accurate. This is because data collection is susceptible to errors. The error could be produced by a mistake on the actual data gathering process or it can manifest itself in the form of gaps or incorrect data input. 

What data cleaning does is to prepare the data for analysis by removing or modifying data that is incorrect, incomplete, irrelevant, duplicated, or improperly formatted. Data cleaning is not about removing incorrect data. By doing this, the accuracy of the data gets compromised because information was removed. Data cleaning's main goal is to maximize the accuracy of the data without necessarily deleting information.

### A. Null Values

Before anything else, checking for **null** values inside the dataset is important because further data cleaning needs to be conducted if there are missing values.

The fastest way to do this is by using the `isnull().values.any()` function call. `isnull` checks each element in the DataFrame, whether if they are null, and returns True if that element is null and False for otherwise. The `values.any()` function speeds up the checking process by allowing the entirety of the DataFrame to be checked and returns True if any element inside the data is null; False for otherwise.

In [5]:
heartfailure_df.isnull().values.any()

False

As observed above, there were no instances of any null elements inside the dataset.

### B. Binary Variables

The next step for data cleaning is to check whether the input for the binary variables from the dataset consists of 0s and 1s only. The binary variables from the dataset are columns `anaemia`, `diabetes`, `high_blood_pressure`, `sex`, `smoking`, and `DEATH_EVENT`.

The `unique` function will be used in this operation for this function returns all the unique elements present. The order of checking is based on the order of the columns in the dataset.

In [6]:
heartfailure_df['anaemia'].unique()

array([0, 1])

In [7]:
heartfailure_df['diabetes'].unique()

array([0, 1])

In [8]:
heartfailure_df['high_blood_pressure'].unique()

array([1, 0])

In [9]:
heartfailure_df['sex'].unique()

array([1, 0])

In [10]:
heartfailure_df['smoking'].unique()

array([0, 1])

In [11]:
heartfailure_df['DEATH_EVENT'].unique()

array([1, 0])

From the execution of the functions above, it can be observed that all the binary variables in the dataset does not include any other value except for 0 and 1. This means that there is no inconsistency in this section of the dataset.

### C. Numerical Values

Checking the consistency of numerical values is a difficult task for there are different criteria for each variable. 

In the given dataset, there is one variable among all the other numerical variables that has to follow a strict rule for its values. The `age` column should only contain whole numbers. Initial inspection of the dataset information indicates that the `age` column has a type of **float64**.

For checking, a class called `HeartFailure` should be imported first.

In [12]:
from heart_failure import HeartFailure

Afterwards, the class will be instantiated with the variable name `hf_func`.

In [13]:
hf_func = HeartFailure()

To check the consistency of the data, a custom function called `check_if_int` was created. As the name states, the function will check whether a column all containes integer data or not. If the results are all True, meaning that all values are integers, then a Boolean value will be returned. Otherwise, the function will return two parameters: the Boolean value and an array list containing the indeces where non-whole number value is found.

In [14]:
print(hf_func.check_if_int(heartfailure_df['age']))

[False, array([185, 188])]


The results above indicate that there are two observations that contain an `age` value that is not a whole number.

To get this row, the result of the function call above will be stored in a variable called `result`. This variable name will be used to access the rows indeces with non-whole number age. To display these rows, the `loc` function will be called.

In [15]:
result = hf_func.check_if_int(heartfailure_df['age'])
heartfailure_df.loc[result[1]]

age  anaemia  creatinine_phosphokinase  diabetes  ejection_fraction  \
185  60.667        1                       104         1                 30   
188  60.667        1                       151         1                 40   

     high_blood_pressure  platelets  serum_creatinine  serum_sodium  sex  \
185                    0   389000.0               1.5           136    1   
188                    1   201000.0               1.0           136    0   

     smoking  time  DEATH_EVENT  
185        0   171            1  
188        0   172            0

Out of 299 observations, there are two obsevations that have the value of 60.667 for the `age`. These two values are also the only instances where an age that is not a whole number is observed. Therefore, the authors concluded that there could have an error or mishandling that occurred during the data collection.

To determine whether the two instances could be dropped or should be replaced, the mean, median, and mode of the `age` column should be acquired.

In [16]:
aggr = heartfailure_df.agg({'age': ['mean', 'median']})
aggr

age
mean    60.833893
median  60.000000

In [17]:
print(heartfailure_df['age'].mode())

0    60.0
dtype: float64


The mean, median, and mode of all the values in `age` column are near each other, along the 60 range. Since data cleaning does not allow explicit alteration of data, the `age` value for the row above cannot be changed. 

Removing the two rows does not dramatically change the shape and range of the dataset. Therefore, the authors deem that action to be viable.

The rows in the dataset can be removed using the `drop` function. By specifying which rows are to be dropped, they can be removed from the dataset itself. Additionally, changing the `inplace` parameter to **True** will directly modify the data frame. This is done so that there is no need to create an entirely new variable for the modified dataset.

In [18]:
heartfailure_df.drop(result[1], inplace=True)

heartfailure_df

age  anaemia  creatinine_phosphokinase  diabetes  ejection_fraction  \
0    75.0        0                       582         0                 20   
1    55.0        0                      7861         0                 38   
2    65.0        0                       146         0                 20   
3    50.0        1                       111         0                 20   
4    65.0        1                       160         1                 20   
..    ...      ...                       ...       ...                ...   
294  62.0        0                        61         1                 38   
295  55.0        0                      1820         0                 38   
296  45.0        0                      2060         1                 60   
297  45.0        0                      2413         0                 38   
298  50.0        0                       196         0                 45   

     high_blood_pressure  platelets  serum_creatinine  serum_sodium  sex  \
0                      1  265000.00               1.9           130    1   
1                      0  263358.03               1.1           136    1   
2                      0  162000.00               1.3           129    1   
3                      0  210000.00               1.9           137    1   
4                      0  327000.00               2.7           116    0   
..                   ...        ...               ...           ...  ...   
294                    1  155000.00               1.1           143    1   
295                    0  270000.00               1.2           139    0   
296                    0  742000.00               0.8           138    0   
297                    0  140000.00               1.4           140    1   
298                    0  395000.00               1.6           136    1   

     smoking  time  DEATH_EVENT  
0          0     4            1  
1          0     6            1  
2          1     7            1  
3          0     7            1  
4          0     8            1  
..       ...   ...          ...  
294        1   270            0  
295        0   271            0  
296        0   278            0  
297        1   280            0  
298        1   285            0  

[297 rows x 13 columns]

### 2.5. (insert header title)

## 3. Research Questions

## 4. Data Modelling

In [64]:
db = heartfailure_df.groupby(["anaemia","DEATH_EVENT"])

list(db

[((0, 0),
        age  anaemia  creatinine_phosphokinase  diabetes  ejection_fraction  \
  23   53.0        0                        63         1                 60   
  38   60.0        0                      2656         1                 30   
  43   72.0        0                       127         1                 50   
  62   55.0        0                       109         0                 35   
  64   45.0        0                       582         0                 80   
  ..    ...      ...                       ...       ...                ...   
  294  62.0        0                        61         1                 38   
  295  55.0        0                      1820         0                 38   
  296  45.0        0                      2060         1                 60   
  297  45.0        0                      2413         0                 38   
  298  50.0        0                       196         0                 45   
  
       high_blood_pressure  platelets  

In [46]:
table = pd.DataFrame([db[0], db[1]], index=["Anemia: 0", "Anemia: 1"])
table

DEATH_EVENT    0   1
Anemia: 0    120  50
Anemia: 1     82  45

## 5. Insights and Conclusion

## 6. References
<pre>
[1] D. Chicco and G. Jurman, "Machine learning can predict survival of patients with heart failure from serum creatinine 
     and ejection fraction alone," BMC Med Inform Decis Mak, vol. 20, no. 1, Feb. 2020, doi: 10.1186/s12911-020-1023-5.  
[2] T. Ahmad, A. Munir, S. H. Bhatti, M. Aftab, and M. A. Raza, "Survival analysis of heart failure patients: A case 
     study," PloS one, vol. 12, no. 7, Jul. 2017, doi: https://doi.org/10.1371/journal.pone.0181001.
</pre>